Tests/Map.cpp

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2023 Magnopus LLC
+
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if !defined(SKIP_INTERNAL_TESTS) || defined(RUN_COMMONTYPE_TESTS) || defined(RUN_COMMONTYPE_MAP_TESTS)
+
+	#include "CSP/Common/Map.h"
+
+	#include "CSP/Common/Optional.h"
+	#include "TestHelpers.h"
+
+	#include <gtest/gtest.h>
+
+using namespace csp::common;
+
+// Test fixture for csp::common::Map testing
+class MapTest
+{
+public:
+	MapTest()
+	{
+		MyMap[1] = "One";
+		MyMap[2] = "Two";
+		MyMap[3] = "Three";
+	};
+
+	Map<int, csp::common::String> MyMap;
+};
+
+// Test case to check if elements are inserted correctly
+CSP_INTERNAL_TEST(CSPEngine, CommonMapTests, MapInsertionTest)
+{
+	auto MapTestLocal = MapTest();
+
+	// Check if the map contains the initial values
+	EXPECT_TRUE(MapTestLocal.MyMap.HasKey(1));
+	EXPECT_TRUE(MapTestLocal.MyMap.HasKey(2));
+	EXPECT_TRUE(MapTestLocal.MyMap.HasKey(3));
+
+	// Check if the map does not contain an invalid key
+	EXPECT_FALSE(MapTestLocal.MyMap.HasKey(4));
+}
+
+// Test case to check if elements are removed correctly
+CSP_INTERNAL_TEST(CSPEngine, CommonMapTests, MapDeletionTest)
+{
+	auto MapTestLocal = MapTest();
+
+	// Remove an element from the map
+	MapTestLocal.MyMap.Remove(2);
+
+	// Check if the removed element is not in the map
+	EXPECT_FALSE(MapTestLocal.MyMap.HasKey(2));
+
+	// Check if the other elements are still in the map
+	EXPECT_TRUE(MapTestLocal.MyMap.HasKey(1));
+	EXPECT_TRUE(MapTestLocal.MyMap.HasKey(3));
+}
+
+// Test case to check if elements are updated correctly
+CSP_INTERNAL_TEST(CSPEngine, CommonMapTests, MapUpdateTest)
+{
+	auto MapTestLocal = MapTest();
+
+	// Update an element in the map
+	MapTestLocal.MyMap[2] = "NewTwo";
+
+	// Check if the updated element has the correct value
+	EXPECT_EQ(MapTestLocal.MyMap[2], "NewTwo");
+}
+
+#endif
```

Comment:
As these are meant to be unit tests, we should be testing every public function of Map with various value types. This means constructors, operator overloads, copy assignment, etc. You can refer to the existing tests for Array, Optional, String, etc.

Reply:
I see that the copy constructor is being tested, but not the move constructor. Copy and move assignment are also missing, as are tests for Keys(), Values(), and Clear().